art.hpp

```diff
@@ -52,21 +91,285 @@ class db final {
+    // If the iterator can not be positioned, it will be set to end().
+    // For example, if [fwd:=true] and the [search_key] is GT any key
+    // in the index then the iterator will be positioned to end()
+    // since there is no index entry greater than the search key.
+    // Likewise, if [fwd:=false] and the [search_key] is LT any key in
+    // the index, then the iterator will be positioned to end() since
+    // there is no index entry LT the search key.
+    //
+    // @param search_key The internal key used to position the iterator.
+    //
+    // @param match Will be set to true iff the search key is an exact
+    // match in the index data.  Otherwise, the match is not exact and
+    // the iterator is positioned either before or after the
+    // search_key.
+    //
+    // @param fwd When true, the iterator will be positioned first
+    // entry which orders GTE the search_key and end() if there is no
+    // such entry.  Otherwise, the iterator will be positioned on the
+    // last key which orders LTE the search_key and end() if there is
+    // no such entry.
+    iterator& seek(const detail::art_key& search_key, bool& match, bool fwd = true) noexcept;
+
+    // Iff the iterator is positioned on an index entry, then returns
+    // the decoded key associated with that index entry.
+    inline std::optional<const key> get_key() noexcept;
+    
+    // Iff the iterator is positioned on an index entry, then returns
+    // the value associated with that index entry.
+    inline std::optional<const value_view> get_val() const noexcept;
+    
+    inline bool operator==(const iterator& other) const noexcept;
+    inline bool operator!=(const iterator& other) const noexcept;
+
+    // Debugging
+    [[gnu::cold]] UNODB_DETAIL_NOINLINE void dump(std::ostream &os) const;
+
+   protected:
+    
+    // Return true unless the stack is empty.
+    inline bool valid() const noexcept { return ! stack_.empty(); }
+
+    // Push the given node onto the stack and traverse from the
+    // caller's node to the left-most leaf under that node, pushing
+    // nodes onto the stack as they are visited.
+    iterator& left_most_traversal(detail::node_ptr node) noexcept;
+
+    // Descend from the current state of the stack to the right most
+    // child leaf, updating the state of the iterator during the
+    // descent.
+    iterator& right_most_traversal(detail::node_ptr node) noexcept;
+
+    // Return the node on the top of the stack, which will wrap
+    // nullptr if the stack is empty.
+    inline detail::node_ptr current_node() noexcept {
+      return stack_.empty()
+          ? detail::node_ptr(nullptr)
+          : std::get<NP>( stack_.top() );
+      ;
+    }
+
+    // Compare the given key (e.g., the toKey) to the current key in
+    // the internal buffer.
+    //
+    // @return -1, 0, or 1 if this key is LT, EQ, or GT the other key.
+    inline int cmp(const detail::art_key& akey) const noexcept {
```

Comment:
TODO belongs to the key, not to the iterator?

Reply:
Moved TODO inside of the method.

It belongs to the iterator too.  I think that iterator needs to maintain a buffer with the materialized key during push and pop operations to support easy key comparisons against external keys and easy materialization of the internal key.